### DISCUSSION TITLE
v4.0.38 pre-release

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#83] [General]
Thanks to @gcoan for adding a sweet Apex Charts PV generation graph example to the readme, along with an explanation of key concepts. I have tipped in some exception handling improvements, along with a nifty spline interpolation for the forecast remaining (which is not earth shattering, but visually pleasing and a better forecast where partial half-hour intervals are involved).

As always, both positive and negative feedback is most welcome, and this will contribute to getting the release ultimately published as latest. Report back, and don't be shy!

Feedback for v4.0.37 was gratefully received from @gcoan, @isorin, @dtswk, @ProphetOfDoom, @EddieSpaghettie and @hwb45, and there are (currently) 124 downloads of that prior release, with no relevant issues raised.

HACS publish is closer. The PR that @BJReplay submitted has progressed to all three required reviews being completed. Merge is the next step, but historically merges for new integrations have been sporadic. (Well, "new" integration in this case is a stretch.) I think February was the last one, with quite a queue outstanding.

> [!NOTE]
> The BJReplay fork is still not published for searching in HACS, so continued use of a manual install in HACS is the go. As explained earlier we're going to keep any releases as pre-release on the fork until that happens, as any 'latest' version change would require modification of that HACS pull request.

## What's Changed
* Add Solcast key concepts and sample PV generation graph to readme by @gcoan
* Add PCHIP spline to forecast remaining by @autoSteve
* Reliability and logging improvements by @autoSteve

**Full Changelog**: https://github.com/BJReplay/ha-solcast-solar/compare/v4.0.37...v4.0.38

**Comment by autoSteve** [comment]
I can't thank everyone enough who are invested in a bright future for this integration (pardon the solar pun).

Feedback and PRs and ideas and just experiences are simply brilliant.

I want this integration to "just work" as much as you do, so bring it! Even if it's a whimpered "yeah, meh, it just works as advertised"...

Solcast are getting feisty at times recently with the 429/back off! messages, and I think we've nailed those with conservative back-off retries. (I monitor every forecast fetch of mine, which is OCD weird, but I do it for you. And none have failed, many retried.) There have recently only been issues raised unrelated to the integration, and I'm sure @BJReplay celebrates those (as I do), whether it ends with "read the readme", or "your networking is broken, but here's how to fix it". I also don't mind being an Agony Aunt for kind-of-related-stuff if I can...

I reckon we're on to a winner with this base, hence the move to actually release new features lately.

I love you all. 🥰 Keep the testing coming, and more importantly the ideas coming. There is only so much one can do with a solar forecast, but maybe you can think of something that no one else has...

If nothing else comes, then roll on HACS merge to become the new default searchable Solcast integration with BJReplay at the helm.

**Comment by autoSteve** [comment]
**Note that multiple beta testers are reporting startup issues with v4.0.38 because the dependency 'scipy' cannot be loaded.**

Many other users are not impacted, which makes me think that this may be platform specific. It is being investigated.

For the time being, downgrade to v4.0.37 should you hit this issue. (Or tell me what the heck is going on...)

**Comment by gcoan** [comment]
Not as important as the issues some are having with Solcast not loading, but I spotted yesterday a remnant of Oziee in the devices list.

Not sure what the extra device is or why its there, and whether it gets resolved when the BJreplay integration gets properly merged in HACS?

Dunno but thought I'd share it:
![image](https://github.com/user-attachments/assets/c5724832-9bef-4d56-8164-4cccfd0f44c8)
![image](https://github.com/user-attachments/assets/6515cd19-3859-4345-b8eb-5b36ab29e93c)

Do others have this or is it just me?

**Reply by autoSteve** [reply]
Huh.

I don't have that.

Is it possible to kill it off using the 'three dots'? The name intrigues, being 'Solcast PV Solar', and not 'Solcast PV Forecast'...

<img width="345" alt="image" src="https://github.com/user-attachments/assets/5a5f371c-eec8-4be7-84cc-d0e8ff6628f6">

**Reply by gcoan** [reply]
> Is it possible to kill it off using the 'three dots'? The name intrigues, being 'Solcast PV Solar', and not 'Solcast PV Forecast'...

the 3 dots only gives a download diagnostics option 

Interesting that @BJReplay has the same two Solcast devices but his second one is owned by byreplay.

Its something to do with the HACS bit of the integration so maybe I will wait until its enrolled properly to HACS and I can uninstall completely and reinstall.  Risk though getting some of the sensor names wrong like I did last time, one of them I slightly misnamed when removing the forecast_forecast so I have the old one orphaned in the statistics and of course new stats for the new name.

**Comment by BJReplay** [comment]
> Do others have this or is it just me?

Just you - looks like a remnant of the oziee install. 

This is what I have under devices when I filter for Solcast... 

Note shitty mid-winter forecasts

![Screenshot_20240715_140829_Home Assistant.jpg](https://github.com/user-attachments/assets/e9867810-3be8-4fe0-809e-5c71d3924839)

![Screenshot_20240715_140847_Home Assistant.jpg](https://github.com/user-attachments/assets/625e5365-d789-4f61-a249-88ac5dd74cae)

![Screenshot_20240715_140903_Home Assistant.jpg](https://github.com/user-attachments/assets/88f4ad76-7028-427a-9ba5-25329fe6763f)

**Reply by gcoan** [reply]
yes really wintery forecasts. Today and yesterday were forecast grey, 40kW for both, but 82 the day after.

Other people further North in the UK have been barely scraping 10kWh so I feel quite lucky about the weather

**Reply by autoSteve** [reply]
My house is solar powered, and I'm sure my body is too. I hate Winter as much as my PowerWalls do. (Shame _I_ can't get a grid charge...)

**Comment by ProphetOfDoom** [comment]
So - Just before taking the leap of faith and upgrading to V4.038, I noticed the following in the system log. Although it suggests rasing an issue with HACS, I thought I'd seek an opinion here first as it mentions solcast_solar.

I have now upgraded to V4.0.38 and nothing untoward to report, though interestingly, and probably unconnected, I've had the first ever 503 error from the Meteorologisk institutt (Met) service, which isn't something I actively use

Logger: homeassistant.util.loop
Source: util/loop.py:136
First occurred: 23:09:27 (1 occurrences)
Last logged: 23:09:27

Detected blocking call to open with args ('/config/custom_components/solcast_solar/recorder.py', 'wb') inside the event loop by custom integration 'hacs' at custom_components/hacs/repositories/base.py, line 600: zip_file.extractall(self.content.path.local) (offender: /usr/local/lib/python3.12/zipfile/__init__.py, line 1801: open(targetpath, "wb") as target:), please create a bug report at https://github.com/hacs/integration/issues For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open Traceback (most recent call last): File "<frozen runpy>", line 198, in _run_module_as_main File "<frozen runpy>", line 88, in _run_code File "/usr/src/homeassistant/homeassistant/__main__.py", line 223, in <module> sys.exit(main()) File "/usr/src/homeassistant/homeassistant/__main__.py", line 209, in main exit_code = runner.run(runtime_conf) File "/usr/src/homeassistant/homeassistant/runner.py", line 190, in run return loop.run_until_complete(setup_and_run_hass(runtime_config)) File "/usr/local/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete self.run_forever() File "/usr/local/lib/python3.12/asyncio/base_events.py", line 641, in run_forever self._run_once() File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1990, in _run_once handle._run() File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run self._context.run(self._callback, *self._args) File "/usr/src/homeassistant/homeassistant/components/websocket_api/decorators.py", line 28, in _handle_async_response await func(hass, connection, msg) File "/config/custom_components/hacs/websocket/repository.py", line 234, in hacs_repository_download await repository.async_install() File "/config/custom_components/hacs/repositories/base.py", line 893, in async_install await self.async_install_repository(version=version) File "/config/custom_components/hacs/repositories/base.py", line 975, in async_install_repository await self.download_zip_files(self.validate) File "/config/custom_components/hacs/repositories/base.py", line 566, in download_zip_files await self.async_download_zip_file( File "/config/custom_components/hacs/repositories/base.py", line 600, in async_download_zip_file zip_file.extractall(self.content.path.local)

**Reply by autoSteve** [reply]
I've not seen that. The integration doesn't unzip anything, so I can't think how it could cause it.

There is a similar blocking issue at GitHub HACS relating to zip_file.extractall (and a duplicate for zip), closed as complete 4 June by ludeus.

**Reply by gcoan** [reply]
I am getting lots of errors from HACS calling things that will be depracated, but not that specific one:

```
Logger: homeassistant.helpers.frame
Source: helpers/frame.py:184
First occurred: 21:14:28 (7 occurrences)
Last logged: 21:14:52

Detected that custom integration 'hacs' calls hass.http.register_static_path which is deprecated because it does blocking I/O in the event loop, instead call `await hass.http.async_register_static_paths([StaticPathConfig("/hacsfiles/iconset.js", "/config/custom_components/hacs/iconset.js", True)])`; This function will be removed in 2025.7 at custom_components/hacs/frontend.py, line 57: hass.http.register_static_path(, please create a bug report at https://github.com/hacs/integration/issues
Detected that custom integration 'hacs' accesses hass.components.frontend. This is deprecated and will stop working in Home Assistant 2024.9, it should be updated to import functions used from frontend directly at custom_components/hacs/frontend.py, line 68: hass.components.frontend.async_register_built_in_panel(, please create a bug report at https://github.com/hacs/integration/issues
Detected that custom integration 'hacs' calls hass.http.register_static_path which is deprecated because it does blocking I/O in the event loop, instead call `await hass.http.async_register_static_paths([StaticPathConfig("/hacsfiles", "/config/www/community", True)])`; This function will be removed in 2025.7 at custom_components/hacs/base.py, line 1162: self.hass.http.register_static_path(, please create a bug report at https://github.com/hacs/integration/issues
Detected that custom integration 'tplink_deco' calls async_forward_entry_setup for integration, tplink_deco with title: 192.168.1.101 and entry_id: 760741b0e8f7b7503e54fb7c71a12b81, which is deprecated and will stop working in Home Assistant 2025.6, await async_forward_entry_setups instead at custom_components/tplink_deco/__init__.py, line 166: config_entry.async_create_task(, please create a bug report at https://github.com/amosyuen/ha-tplink-deco/issues
Detected that custom integration 'hacs' accesses hass.helpers.event. This is deprecated and will stop working in Home Assistant 2024.11, it should be updated to import functions used from event directly at custom_components/hacs/base.py, line 630: self.hass.helpers.event.async_track_time_interval(, please create a bug report at https://github.com/hacs/integration/issues
```

**Reply by ProphetOfDoom** [reply]
@autoSteve Looking at the timestamps on the HA files (23:09:27), it looks like I may have got my timeline wrong and this error/warning came about when performing the update to V4.0.38. Would this explain the attempted unzipping operation?

![image](https://github.com/user-attachments/assets/866637ec-d717-4fd4-9553-904bbbaf69f9)

**Reply by ProphetOfDoom** [reply]
OR - it came about when I logged in to Home Assistant and HACS presented me with the update dialog for V4.0.38. Is this a more likely suspect?

**Reply by ProphetOfDoom** [reply]
@autoSteve I see the closed issue on GitHub HACS. Guess we'll just have to wait for a new HACS release, as the last one (1.34.0) was back in January

**Comment by ProphetOfDoom** [comment]
Could I suggest the following clarification in bold to the documentation section early on regarding api poll limits. I know it's mentioned later on, but users have a habit of not reading to the end!

Note

Solcast have altered their API limits for new account creators

Solcast now only offer new account creators 10 API calls per day (used to be 50). Old account users still have 50 API calls.

The integration currently no longer includes auto API polling. Users now need to create their own automations to call the update solcast service to poll for new data. Keep in mind your API poll limit, **especially if you have more than one roof definition**

**Reply by autoSteve** [reply]
Doh! Suppress the ADHD and read to the end.

I should embed a little easter egg in there somewhere near the bottom that says something like, "If you find this text I owe you a beer"...

**Reply by gcoan** [reply]
I wrote a long and technical outline IT system design for a customer that spread to 100+ pages long.  Towards the end when explaining one of the data interfaces I said that this interface would solve the middle East crisis and bring about world peace.
They never came back to me with any comments. Showed they never read it

**Reply by autoSteve** [reply]
Similar story about a Government tender I worked on. On about page 500 was the offer of a significant discount should the customer provide X. We won. They never offered up X, nor an explanation or mention, so we responded in kind with the original tendered price...

**Reply by BJReplay** [reply]
Two similar stories:

Year 12 physics - I was the physics teacher's favourite student because I got physics - it just made sense. So, for my option, I did "The physics of low frequency sound propagation" - in other words could I turn mucking around with some serious power amplification and W bins and electric bass guitars, and for shits and giggles sound meters, oscilloscopes into a physics option paper? Turned out I not only could, but could embed a "please indicate if you believe that this was a) the sound meter could not detect the sub-harmonic or b) an issue with inductance in the oscilloscope leads" A/B choice as the final sentence in the conclusion with two boxes A or B to tick, and my report was handed back to me with "10/10 excellent report" written immediately below that.

After I got my final mark, I asked my teacher why he did not choose which option he thought most likely - he didn't know what I was asking about so I showed him. He admitted that he only read the first few pages!

Second story - work - we had a contract where the customer had a standard termination for convenience clause.  We knew that the customer was likely to go through a change of control - M&A was rife in the industry. We added a variation that still allowed termination for convenience, but provided that we had invoiced a project stage, or notified the customer that we were about to commence a project stage in not less than four weeks, then the fee for the invoiced or notified stage was payable.

It was no surprise to us that the customer was sold. We saw it coming (strangers walking the floor) so managed to notify commencement of the final project stage. The project was cancelled. 

The old management, new owners, and no doubt the smart young things who did due diligence who missed our right were surprised, annoyed but realised that they had to pay us out.

Read the fine print, people!

**Comment by gcoan** [comment]
I feel I am truly down Alice's rabbit hole ...

We had a good conversation about units of measurement on the last release, and I thought I understood them now, but one thing we didn't discuss was the pv_today detailed forecasts.
I thought these were straightforward being the kWh for the time period, but having stared at them for a while I am with that white rabbit confused.

Vis, the 30 minute forecast for today, 15:00 is 4.3955 and 15:30 is 4.1503 so to my simple mind each of these is a forecast of energy generated in the half hour period, so add them together and its 8.6458kWh in total over the entire hour:
![image](https://github.com/user-attachments/assets/04f8715d-0fd1-492d-8a90-26b229395b83) 

but yet the hourly forecast says its just 4kWh:
![image](https://github.com/user-attachments/assets/1af1c578-aa09-4983-89d0-0640728a7eca)

Predbat (which is where I started this whole questioning from) has the half hour forecasts as precisely half the Solcast attribute values so suggests it knows what is going on more than I do:
![image](https://github.com/user-attachments/assets/38482e85-f503-4dc2-9707-cd3ef3c8ad2a)

So is this a really big and glaring bug in the detailed half hourly forecasts, is it that a clear explanation of what is actually being presented is needed, is Predbat right or wrong and do I need to go for breakfast in the hope that more sugar will help me see things more clearly?

**Reply by autoSteve** [reply]
When adding half-hourly forecasts (which are really average expected production for each half hour period), add them, then divide by two to get an approximated kWh. Or just choose any half-hour period and divide that by two, because it's expected production for half-a-kWh.

Not see issue. Predbat smart. (_Predbat looks really cool..._) You, more sugar, although it's around morning tea time for you. So have some more sugar with your tea...

I shall get some sugar from a beer.

**Comment by autoSteve** [comment]
In case you've missed it, I've started a discussion about v4.0.39-alpha.

This carries a fix for the scipy woes on some platforms (which wasn't really worth an alpha, just math), but then for the really alpha-ish stuff, it also carries an expanded integration configuration, allowing a reduction in the number of attributes that are published for each sensor.
